### PR TITLE
[verifying-client] Refactor Storage->StateStore. Add FileStateStore.

### PR DESCRIPTION
**Pull request by phlip9**
+ Flesh out the durable `TrustedState` storage API. Replaces the `Storage`+`TrustedStateStore` with new `StateStore` trait. Looks like:

```rust
pub trait StateStore {
    type Error: std::error::Error + Send + Sync + 'static;

    fn latest_state(&self) -> Result<Option<TrustedState>, Self::Error>;
    fn latest_state_version(&self) -> Result<Option<u64>, Self::Error> {
        Ok(self.latest_state()?.map(|s| s.version()))
    }
    fn store(&self, new_state: &TrustedState) -> Result<(), Self::Error>;
}
```

+ Clarify properties needed from `StateStore` implementations in order to guarantee client safety under concurrency and crashes.

+ Add `FileStateStore` which dumps states into one of a pair of files. The main idea here is to provide ACID guarantees by ensuring (1) only one writer, (2) writer always writes to the oldest state file (by version), (3) we fsync the file after every write. This ensures the highest version state is always durable and consistent.

### CODE REVIEW

File: sdk/client/src/verifying_client/client.rs [hunk 1]
```diff
@@ -60,49 +53,53 @@ use std::{
...
-#[derive(Debug)]
+#[derive(Clone, Debug)]
 pub struct VerifyingClient<S> {
     inner: Client,
-    trusted_state_store: Arc<RwLock<TrustedStateStore<S>>>,
+    state_store: S,
 }
 
-impl<S: Storage> VerifyingClient<S> {
+impl<S: StateStore + Clone + Debug> VerifyingClient<S> {
```
[the remark this answers is not in the document]
Comment (phlip9) [reply sdk/client/src/verifying_client/client.rs]: ack forgot to take this out after debugging...

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A void 1.0.2 (direct third-party, crates.io)
    * features: default, std
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A void 1.0.2 (direct third-party, crates.io)
    * features: default, std
```

File: sdk/client/src/verifying_client/state_store.rs [hunk 2]
```diff
@@ -1,110 +1,202 @@
...
-
-mod private {
-    pub trait Sealed {}
+use crate::error::Result;
+use diem_types::trusted_state::TrustedState;
+use std::sync::{Arc, RwLock};
+
+// TODO(philiphayes): how to handle errors in a way that users can implement their
+// own state store backends? do we need to make crate::Error constructors pub?
+// add a new error type?
```
[the remark this answers is not in the document]
Comment (phlip9) [reply sdk/client/src/verifying_client/state_store.rs]: done! I've decided to let each StateStore handle it's own retries internally, which makes this simpler

File: sdk/client/src/verifying_client/file_state_store.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        self.version = Some(new_state.version());
+        Ok(())
+    }
+}
+
+///////////
+// Utils //
+///////////
+
+/// TODO(philiphayes): use std::cmp::max_by_key when it's stable
```
[the remark this answers is not in the document]
Comment (phlip9) [reply sdk/client/src/verifying_client/file_state_store.rs]: just in time! : p

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A void 1.0.2 (direct third-party, crates.io)
    * features: default, std
```

File: sdk/client/Cargo.toml [hunk 4]
```diff
@@ -23,6 +23,7 @@ hex = "0.4.3"
 serde = { version = "1.0.124", features = ["derive"] }
 serde_json = "1.0.64"
 tracing = "0.1.26"
+void = "1.0.2"
```
Comment (bmwill) [inline sdk/client/Cargo.toml:26]: Remove this dependency and use [`Infallible`](https://doc.rust-lang.org/stable/std/convert/enum.Infallible.html) instead maybe?

Comment (phlip9) [reply sdk/client/Cargo.toml:26]: Oh wow, perfect. didn't even realize that existed

**Review by bmwill — DISMISSED** [review]

**Pull request merged**